RevenueCatUI/Runtime/Platforms/iOS/IOSPaywallPresenter.cs

```diff
@@ -0,0 +1,45 @@
+#if UNITY_IOS && !UNITY_EDITOR
+using System;
+using System.Runtime.InteropServices;
+using System.Threading.Tasks;
+
+namespace RevenueCat.UI.Platforms
+{
+    internal class IOSPaywallPresenter : IPaywallPresenter
+    {
+        private delegate void PaywallResultCallback(string result);
+
+        [DllImport("__Internal")] private static extern void rcui_presentPaywall(string offeringIdentifier, bool displayCloseButton, PaywallResultCallback cb);
+        [DllImport("__Internal")] private static extern void rcui_presentPaywallIfNeeded(string requiredEntitlementIdentifier, string offeringIdentifier, bool displayCloseButton, PaywallResultCallback cb);
+        [DllImport("__Internal")] private static extern bool rcui_isSupported();
+
+        private static TaskCompletionSource<PaywallResult> s_current;
+
+        public bool IsSupported() => rcui_isSupported();
+
+        public Task<PaywallResult> PresentPaywallAsync(PaywallOptions options)
+        {
+            s_current = new TaskCompletionSource<PaywallResult>();
+            rcui_presentPaywall(options?.OfferingIdentifier, options?.DisplayCloseButton ?? false, OnResult);
+            return s_current.Task;
+        }
+
+        public Task<PaywallResult> PresentPaywallIfNeededAsync(string requiredEntitlementIdentifier, PaywallOptions options)
+        {
+            s_current = new TaskCompletionSource<PaywallResult>();
+            rcui_presentPaywallIfNeeded(requiredEntitlementIdentifier, options?.OfferingIdentifier, options?.DisplayCloseButton ?? false, OnResult);
+            return s_current.Task;
+        }
+
+        [AOT.MonoPInvokeCallback(typeof(PaywallResultCallback))]
+        private static void OnResult(string result)
+        {
+            var token = (result ?? "ERROR");
+            var native = token.Split('|')[0];
+            var type = PaywallResultTypeExtensions.FromNativeString(native);
+            s_current?.TrySetResult(new PaywallResult(type));
```

Comment:
Hmm I'm curious how `s_current` gets here TBH 😅 Is this some sort of magic Unity provides to get the current task? 🪄 . I see the `[AOT.MonoPInvokeCallback(typeof(PaywallResultCallback))]` but not entirely sure how it works... could you explain a bit?

Reply:
1. In PresentPaywallAsync we create a TaskCompletionSource (s_current) and call into native with a function pointer
  to OnResult. (see https://github.com/RevenueCat/purchases-unity/pull/664/files/0a407d2092f85d88ecafb6aedb6d107b53d6091f#diff-cd801b330a4ad2c0654773bcec01ebe96dfdff0cd1f592c55a2ebdb6f96a242fR22)
2. When native finishes, it calls that function with the result string.
3. OnResult reads s_current, parses the result, sets the Task’s result, then clears s_current.